monkey/infection_monkey/island_api_client/http_requests_facade.py

```diff
@@ -0,0 +1,88 @@
+import functools
+import logging
+from enum import Enum, auto
+
+import requests
+from requests.adapters import HTTPAdapter
+from urllib3 import Retry
+
+from common.common_consts.timeouts import MEDIUM_REQUEST_TIMEOUT
+
+from .island_api_client_errors import (
+    IslandAPIConnectionError,
+    IslandAPIError,
+    IslandAPIRequestError,
+    IslandAPIRequestFailedError,
+    IslandAPITimeoutError,
+)
+
+logger = logging.getLogger(__name__)
+
+
+# Retries improve reliability and slightly mitigate performance issues
+RETRIES = 5
+
+
+class RequestMethod(Enum):
+    GET = auto()
+    POST = auto()
+    PUT = auto()
+
+
+class HTTPRequestsFacade:
+    def __init__(self, api_url: str, retries=RETRIES):
+        self._session = requests.Session()
+        retry_config = Retry(retries)
+        self._session.mount("https://", HTTPAdapter(max_retries=retry_config))
+        self._api_url = api_url
+
+    def get(self, *args, **kwargs) -> requests.Response:
+        return self._send_request(RequestMethod.GET, *args, **kwargs)
+
+    def post(self, *args, **kwargs) -> requests.Response:
+        return self._send_request(RequestMethod.POST, *args, **kwargs)
+
+    def put(self, *args, **kwargs) -> requests.Response:
+        return self._send_request(RequestMethod.PUT, *args, **kwargs)
+
+    def _send_request(
+        self,
+        request_type: RequestMethod,
+        endpoint: str,
+        timeout=MEDIUM_REQUEST_TIMEOUT,
+        data=None,
+        *args,
+        **kwargs,
+    ) -> requests.Response:
+        url = f"{self._api_url}/{endpoint}".strip("/")
+        logger.debug(f"{request_type.name} {url}, timeout={timeout}")
+
+        method = getattr(self._session, str.lower(request_type.name))
+        response = method(url, *args, timeout=timeout, verify=False, json=data, **kwargs)
+        response.raise_for_status()
+
+        return response
+
+
+def handle_island_errors(fn):
```

Comment:
This is only used in `http_island_api_client.py`. Why is it here?

Reply:
Because it handles `requests` package specific errors and returns custom island errors. This is `requests` facade, so all code that interfaces with this library should be here.

Reply:
It's still a decorator to handle errors in the Island API client.

If the argument is that this code is related to `requests`, the `requests` errors should be handled here as well.

Reply:
Only decorating `_send_response` makes a lot of sense, thanks